src/Language/SystemVerilog/Parser/Preprocess.hs

```diff
@@ -141,13 +141,46 @@ getBuffer = do
     p <- getPosition
     return (x, p)
 
+-- Push a condition onto the top of the preprocessor condition stack
+pushCondStack :: Cond -> PPS ()
+pushCondStack c = getCondStack >>= setCondStack . (c :)
+
+-- Pop the top from the preprocessor condition stack
+popCondStack :: String -> PPS Cond
+popCondStack directive = do
+    cs <- getCondStack
+    case cs of
+        [] -> lexicalError $
+              "`" ++ directive ++ " directive outside of an `if/`endif block"
+        c : cs' -> setCondStack cs' >> return c
+
 isIdentChar :: Char -> Bool
 isIdentChar ch =
     ('a' <= ch && ch <= 'z') ||
     ('A' <= ch && ch <= 'Z') ||
     ('0' <= ch && ch <= '9') ||
     (ch == '_') || (ch == '$')
 
+-- Update a Cond for an `else block: if we've never had a hit, this is the one
+-- for us. Otherwise a previous block was true, so it isn't.
+elseCond :: Cond -> Cond
+elseCond NeverTrue = CurrentlyTrue
+elseCond _ = NeverTrue
+
+-- Generate a Cond from a boolean. This is supposed to correspond to the first
+-- test in an if/else chain (so false goes to NeverTrue).
+boolToCond :: Bool -> Cond
+boolToCond True = CurrentlyTrue
+boolToCond False = NeverTrue
+
+-- Update a Cond for an `elsif block. The boolean argument is whether the
+-- `elsif block's test is true.
+elsifCond :: Bool -> Cond -> Cond
+elsifCond defined c =
+    case c of
+        NeverTrue -> boolToCond defined
+        _ -> PreviouslyTrue
+
```

Comment:
I liked how you previously had this set of functions (`elseCond `, `boolToCond`, `elsifCond`) defined right after the `data Cond`.

Reply:
I don't think they've moved? But I agree that's a good place for them: I've moved them up in the latest version.